tower-util/src/combinators/with.rs

```diff
@@ -0,0 +1,70 @@
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub struct With<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> With<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        With { inner, f }
+    }
+}
+
+impl<S, F, NewRequest, OldRequest> Service<NewRequest> for With<S, F>
```

Comment:
Currently, `With` service only support infallible mapping of the request to the new request - that is, `NewRequest -> OldRequest`.

This could be extended to, or complimented with a `TryWith`(?) service where `NewRequest -> Result<OldRequest, Error>` where `Error` is either `Into<ServiceError>` or `Error=ServiceError`.

Reply:
I don't think there's an easy way to have this service implement both fallible and infallible mapping, so adding a separate `TryWith` sounds good to me.

Reply:
It's added with `type Future = Either<S::Service, Ready<Result<S::Response, S::Error>>>`.